### PR TITLE
Add examples for SYCL Spark

### PR DESCRIPTION
**Pull request by jasonsewall-intel**
We should consider this a draft, since I'd like your comments. It's useful to have this out here to enable discussion.

Most noteworthy is that the histogram example doesn't compile with the current intel/llvm SYCL implementation because we do not support array reductions (but see https://github.com/intel/llvm/issues/5941)

### CODE REVIEW

File: src/sycl/histogram.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+double histogram::run() {
+#if defined(__INTEL_LLVM_COMPILER)
+#warning("histogram not supported on Intel/llvm compiler")
+#else
+  pdata->q.submit([&](sycl::handler &h) {
+    int *A = pdata->A;
+    h.parallel_for(
+      sycl::range<1>(N),
+      sycl::reduction(sycl::span<int>(pdata->histogram,16), std::plus<>()),
```
Comment (Pennycook) [inline src/sycl/histogram.cpp:65]:
```suggestion
      sycl::reduction(sycl::span<int, 16>(pdata->histogram,16), std::plus<>()),
```

The second template argument defaults to `dynamic_extent`, which we don't want.

**Review by Pennycook — CHANGES_REQUESTED** [review]

**Comment by Pennycook** [comment]
With the change from https://github.com/UoB-HPC/everythingsreduced/pull/16#discussion_r851307803, the code compiles using the functionality from https://github.com/intel/llvm/commit/863383bac8eeab168f39f4da8e94df860458cbf0.

It doesn't pass the correctness test, but I think that's because nothing here resets the histogram.  The test expects `histogram[8] == N`, but the behavior I see is that `histogram[8] == N * iteration`.  Adding `property::reduction::initialize_to_identity{}` fixes it, but I don't understand why this is necessary -- some of the other benchmarks have this, but not all of them.

@tomdeakin, could you please take a look?  What do you think the correct fix is, here?

**Comment by tomdeakin** [comment]
@Pennycook All the original benchmarks set `property::reduction::initialize_to_identity{}` so if it's missing it's likely an oversight in the new benchmarks. That being said, it's up the benchmark to decide how it tests correctness, so if we want to preserve the initial value, that's OK too - it just needs a comment in the header.

File: src/sycl/histogram.cpp (hunk 1 above)
Comment (Pennycook) [inline src/sycl/histogram.cpp:65]:
```suggestion
      sycl::reduction(sycl::span<int>(pdata->histogram,16), std::plus<>(), sycl::property::reduction::initialize_to_identity{}),
```

**Comment by Pennycook** [comment]
Thanks, @tomdeakin.  I don't think I have permissions to modify Jason's branch, and I think he's lost access to his "-intel" GitHub account.... So, I've just added a suggested change to the line that I think needs to be updated.

Do you have permissions to accept the suggested changes?

File: src/sycl/histogram.cpp (hunk 1 above)
Comment (tomdeakin) [inline src/sycl/histogram.cpp]:
```suggestion
      sycl::reduction(sycl::span<int, 16>(pdata->histogram,16), std::plus<>()),
```
GitHub won't let me accept an Outdated suggestion, so re-implementing and accepting it.

File: src/sycl/histogram.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+double histogram::run() {
+#if defined(__INTEL_LLVM_COMPILER)
+#warning("histogram not supported on Intel/llvm compiler")
+#else
+  pdata->q.submit([&](sycl::handler &h) {
+    int *A = pdata->A;
+    h.parallel_for(
+      sycl::range<1>(N),
+      sycl::reduction(sycl::span<int>(pdata->histogram,16), std::plus<>(), sycl::property::reduction::initialize_to_identity{}),
```
Comment (tomdeakin) [inline src/sycl/histogram.cpp:65]:
```suggestion
      sycl::reduction(sycl::span<int, 16>(pdata->histogram,16), std::plus<>(), sycl::property::reduction::initialize_to_identity{}),
```

**Comment by tomdeakin** [comment]
@Pennycook - think I have those changes in now.

We might want to merge this and #17 into a `spark` branch in the short term as both these PRs will likely break compiling the other models which don't implement these kernels (yet).

**Comment by jasonsewall-intel** [comment]
I still have access! I changed the email address before I left. But this looks good and I thank you for making the changes.

**Comment by tomdeakin** [comment]
All good to merge into a `spark` branch? I can then merge #17 into the same branch; Im predicting minor conflicts in the driver code but should be simple to fix.

**Comment by Pennycook** [comment]
> All good to merge into a `spark` branch? I can then merge #17 into the same branch; Im predicting minor conflicts in the driver code but should be simple to fix.

Sounds good to me.  Thanks again for doing this.  And sorry, @jasonsewall-intel -- I don't know why I thought you'd lost access!

**Pull request merged**